### PR TITLE
Random Fourier Feature Kernel Approximation

### PR DESCRIPTION
**Pull request by thomaspinder**
This PR introduces a Random Fourier Feature approximation for the RBF and Matérn kernels. 

## Pull request type

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

There are currently no implemented kernel approximations in JaxKern.

Issue Number: #39 

## Other information

No example notebooks provided as we're moving towards a centralised set of docs.

### CODE REVIEW

File: jaxkern/approximations/rff.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        base_params = self.base_kernel.init_params(key)
+        n_dims = base_params["lengthscale"].shape[0]
+        frequencies = self.base_kernel.spectral_density.sample(
+            seed=key, sample_shape=(self.num_basis_fns, n_dims)
+        )
+        base_params["frequencies"] = frequencies
+        return base_params
+
+    def __call__(self, *args: Any, **kwds: Any) -> Any:
+        pass
```
Comment (henrymoss) [inline jaxkern/approximations/rff.py:27]: Perhaps stupid q,  so call is never called by the user really? It just feeds into cross_covariance?

File: jaxkern/approximations/rff.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from ..base import AbstractKernel
+from ..computations import BasisFunctionComputation
+from jax.random import KeyArray
+from typing import Dict, Any
+
+
+class RFF(AbstractKernel):
+    def __init__(self, base_kernel: AbstractKernel, num_basis_fns: int) -> None:
+        self._check_valid_base_kernel(base_kernel)
```
Comment (henrymoss) [inline jaxkern/approximations/rff.py:9]: Are we deliberately light on doc strings?

File: jaxkern/approximations/rff.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        self.base_kernel = base_kernel
+        self.num_basis_fns = num_basis_fns
+        # Set the computation engine to be basis function computation engine
+        self.compute_engine = BasisFunctionComputation
+        # Inform the compute engine of the number of basis functions
+        self.compute_engine.num_basis_fns = num_basis_fns
+
+    def init_params(self, key: KeyArray) -> Dict:
+        base_params = self.base_kernel.init_params(key)
+        n_dims = base_params["lengthscale"].shape[0]
```
Comment (henrymoss) [inline jaxkern/approximations/rff.py:19]: If we ever do spectral decomp for a kernel without "lengthscale" then we are in trouble.

File: tests/test_approximations.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    better_approximation = RFF(base_kernel, num_basis_fns=50)
+    b_params = better_approximation.init_params(key)
+    b_linop = better_approximation.gram(b_params, x).to_dense()
+
+    c_delta = jnp.linalg.norm(exact_linop - c_linop, ord="fro")
+    b_delta = jnp.linalg.norm(exact_linop - b_linop, ord="fro")
+
+    # The frobenius norm of the difference between the exact and approximate
+    # should improve as we increase the number of basis functions
+    assert c_delta > b_delta
```
Comment (henrymoss) [inline tests/test_approximations.py:123]: should we do a test where we check that c_delta and b_delta are close ?

File: tests/test_approximations.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+@pytest.mark.parametrize("kernel", [RBF, Matern12, Matern32, Matern52])
+@pytest.mark.parametrize("num_basis_fns", [2, 10, 20])
+@pytest.mark.parametrize("n_dims", [1, 2, 5])
+def test_frequency_sampler(kernel: AbstractKernel, num_basis_fns: int, n_dims: int):
+    key = jr.PRNGKey(123)
+    base_kernel = kernel(active_dims=list(range(n_dims)))
+    approximate = RFF(base_kernel, num_basis_fns)
+
+    params = approximate.init_params(key)
+    assert params["frequencies"].shape == (num_basis_fns, n_dims)
```
Comment (henrymoss) [inline tests/test_approximations.py:34]: At no point do we test that the right spectral density is coming out of the kernels. Should we?

i.e. rbf -> Normal e.t.c.

**Review by henrymoss — APPROVED** [review]
Had a go

File: jaxkern/stationary/matern32.py [hunk 6]
```diff
@@ -21,24 +21,24 @@
...
 class Matern32(AbstractKernel):
     """The Matérn kernel with smoothness parameter fixed at 1.5."""
 
     def __init__(
         self,
-        compute_engine: AbstractKernelComputation = DenseKernelComputation,
         active_dims: Optional[List[int]] = None,
-        stationary: Optional[bool] = False,
-        spectral: Optional[bool] = False,
+        stationary: Optional[bool] = True,
```
[the remark this answers is not in the document]
Comment (henrymoss) [reply jaxkern/stationary/matern32.py:35]: i.e. why is stationary somethig we can control?

File: jaxkern/computations/constant_diagonal.py [hunk 7]
```diff
@@ -51,10 +51,11 @@ def gram(
         Returns:
             CovarianceOperator: The computed square Gram matrix.
         """
-
         value = self.kernel_fn(params, inputs[0], inputs[0])
 
-        return ConstantDiagonalLinearOperator(value = jnp.atleast_1d(value), size=inputs.shape[0])
+        return ConstantDiagonalLinearOperator(
+            value=jnp.atleast_1d(value), size=inputs.shape[0]
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/computations/constant_diagonal.py:57]: Everything is Black formatted?

File: jaxkern/stationary/matern32.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/stationary/matern32.py:35]: Resolved.

File: jaxkern/computations/constant_diagonal.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (henrymoss) [reply jaxkern/computations/constant_diagonal.py:57]: this bit wasnt before?

File: jaxkern/computations/basis_functions.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+import jax.numpy as jnp
+from jaxtyping import Array, Float
+from .base import AbstractKernelComputation
+from jaxlinop import DenseLinearOperator
+
+
+class BasisFunctionComputation(AbstractKernelComputation):
+    """Compute engine class for finite basis function approximations to a kernel."""
+
```
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/computations/basis_functions.py:11]: Opened an issue #45

File: jaxkern/approximations/rff.py (hunk 1 above)
Comment (thomaspinder) [reply jaxkern/approximations/rff.py:27]: In this case, no. In general, `__call__` is the evaluation of $k(x_i, x_j)$.

File: jaxkern/computations/constant_diagonal.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (thomaspinder) [reply jaxkern/computations/constant_diagonal.py:57]: That’s odd - it should be as black is in the pre-commits.

File: jaxkern/approximations/rff.py (hunk 2 above)
Comment (thomaspinder) [reply jaxkern/approximations/rff.py:9]: No - sloppy PR. Resolved.

File: jaxkern/approximations/rff.py (hunk 3 above)
Comment (thomaspinder) [reply jaxkern/approximations/rff.py:19]: How likely is that? Personally, I’m OK with it for now. The workaround would be adding an attribute to each kernel that tells us whether or not it is parameterised by a lengthscale. If no, then a vector of ones would be created in the RFF approximation. To me, this feels a little messy.

File: tests/test_approximations.py (hunk 5 above)
Comment (thomaspinder) [reply tests/test_approximations.py:34]: Good idea - added as a test.

File: tests/test_approximations.py (hunk 4 above)
Comment (thomaspinder) [reply tests/test_approximations.py:123]: Yes - resolved.

**Comment by thomaspinder** [comment]
All PR comments have now been actioned @henrymoss

File: tests/test_approximations.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    exact_params = kernel.init_params(key)
+    exact_linop = kernel.gram(exact_params, x).to_dense()
+
+    better_approximation = RFF(kernel, num_basis_fns=500)
+    b_params = better_approximation.init_params(key)
+    b_linop = better_approximation.gram(b_params, x).to_dense()
+
+    max_delta = jnp.max(exact_linop - b_linop)
+    assert max_delta < 0.1
+
```
Comment (henrymoss) [inline tests/test_approximations.py:141]: looks good. But I have no idea if 0.1 is a tight tolerance or not

**Review by henrymoss — APPROVED** [review]
LGTM. I think the maths is right but a test with a tighter tolerance would make me happier

**Pull request merged**